packages/ra-core/src/dataProvider/useUpdate.ts

```diff
@@ -45,74 +55,319 @@ import useMutation, { MutationOptions, Mutation } from './useMutation';
+                    return reactMutationOptions.onSettled(
+                        data,
+                        error,
+                        variables,
+                        context
+                    );
+                }
+            },
+        }
     );
 
-    return [update, state];
+    const mutateAsync = async (
+        variables: Partial<UseUpdateParams<RecordType>> = {},
+        callTimeOptions: MutateOptions<
+            RecordType,
+            unknown,
+            Partial<UseUpdateParams<RecordType>>,
+            unknown
+        > = {}
+    ) => {
+        if (mutationMode === 'pessimistic') {
+            return mutation.mutateAsync(variables, callTimeOptions);
+        }
+
+        const { onSuccess, onSettled, onError } = callTimeOptions;
+        const {
+            resource: callTimeResource = resource,
+            id: callTimeId = id,
+            data: callTimeData = data,
+        } = variables;
+
+        // optimistic update as documented in https://react-query.tanstack.com/guides/optimistic-updates
+        // except we wo it in a mutate wrapper instead of the onMutete callback
+        // to have access to success side effects
+
+        // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+        await queryClient.cancelQueries([
+            callTimeResource,
+            'getOne',
+            callTimeId,
+        ]);
+        await queryClient.cancelQueries([resource, 'getList']);
+
+        // Snapshot the previous values
+        const previousGetOne = queryClient.getQueryData([
+            callTimeResource,
+            'getOne',
+            callTimeId,
+        ]);
+        const previousGetList = queryClient.getQueriesData([
+            callTimeResource,
+            'getList',
+        ]);
+
+        // Optimistically update to the new value in getOne
+        await updateCache({
+            resource: callTimeResource,
+            id: callTimeId,
+            data: callTimeData,
+        });
+
+        context = { previousGetOne, previousGetList };
+
+        // run the success callbacks during the next tick
+        if (onSuccess) {
+            setTimeout(
+                () =>
+                    onSuccess(
+                        queryClient.getQueryData([
+                            callTimeResource,
+                            'getOne',
+                            callTimeId,
+                        ]),
+                        variables,
+                        context
+                    ),
+                0
+            );
+        }
+
+        if (mutationMode === 'optimistic') {
+            // call the mutate without success side effects
+            return mutation.mutateAsync(variables, {
+                onSettled,
+                onError,
+            });
+        } else {
+            // undoable mutation: register the mutation for later
+            undoableEventEmitter.once('end', ({ isUndo }) => {
```

Comment:
How can we be sure this event was dispatched for this specific query? What if there are multiple queries sent in a short timeframe?

Reply:
We can't. That's the way it'as always worked. 

But it's not a real problem as, in order to perform a new mutation, users have to click somewhere, which dismisses the MUI Notification and commits the initial update.

Reply:
Ok. That's only true for our implementation of the ui though